pkg/controllers/networking/pod_controller.go

```diff
@@ -183,61 +184,57 @@ func (r *PodReconciler) selectNetwork(pod *corev1.Pod) (string, error) {
 		// try to get underlay network by node indexer
-		var networkList *networkingv1.NetworkList
-		var err error
-		if networkList, err = utils.ListNetworks(r, client.MatchingFields{IndexerFieldNode: pod.Spec.NodeName}); err != nil {
-			return "", fmt.Errorf("unable to list underlay network by indexer node: %v", err)
-		}
-		if len(networkList.Items) >= 1 {
-			return networkList.Items[0].GetName(), nil
+		underlayNetworkName, err := r.getNetworkByNodeNameIndexer(pod.Spec.NodeName)
+		if err != nil {
+			return "", fmt.Errorf("unable to get underlay network by node name indexer: %v", err)
 		}
 
-		// fall back to find underlay network by label selector
-		var underlayNetworkName string
-		if underlayNetworkName, err = utils.FindUnderlayNetworkForNodeName(r, pod.Spec.NodeName); err != nil {
-			return "", fmt.Errorf("unable to find underlay network for node %s", pod.Spec.NodeName)
-		}
-		if len(underlayNetworkName) == 0 {
-			return "", fmt.Errorf("no underlay network match node %s", pod.Spec.NodeName)
-		}
-		if !r.matchNetworkTypeInManager(underlayNetworkName, types.Underlay) {
-			return "", fmt.Errorf("network %s does not match type %q in manager", underlayNetworkName, types.Underlay)
+		if len(underlayNetworkName) != 0 {
+			return underlayNetworkName, nil
 		}
-		return underlayNetworkName, nil
+
+		return "", fmt.Errorf("unable to find underlay network for node %s", pod.Spec.NodeName)
 	case types.Overlay:
 		// try to get overlay network by special node name
-		var networkList *networkingv1.NetworkList
-		var err error
-		if networkList, err = utils.ListNetworks(r, client.MatchingFields{IndexerFieldNode: OverlayNodeName}); err != nil {
-			return "", fmt.Errorf("unable to list overlay network by indexer node: %v", err)
-		}
-		if len(networkList.Items) >= 1 {
-			return networkList.Items[0].GetName(), nil
+		overlayNetworkName, err := r.getNetworkByNodeNameIndexer(OverlayNodeName)
+		if err != nil {
+			return "", fmt.Errorf("unable to get overlay network by node name indexer: %v", err)
 		}
 
-		// fall back to find overlay network in client cache
-		var overlayNetworkName string
-		if overlayNetworkName, err = utils.FindOverlayNetwork(r); err != nil {
-			return "", fmt.Errorf("unable to find overlay network")
+		if len(overlayNetworkName) != 0 {
+			return overlayNetworkName, nil
 		}
-		if len(overlayNetworkName) == 0 {
-			return "", fmt.Errorf("no overlay network found")
+
+		return "", fmt.Errorf("unable to find overlay network")
+	case types.GlobalBGP:
+		// try to get global bgp network by special node name
+		globalBGPNetworkName, err := r.getNetworkByNodeNameIndexer(GlobalBGPNodeName)
+		if err != nil {
+			return "", fmt.Errorf("unable to get overlay network by node name indexer: %v", err)
 		}
-		if !r.matchNetworkTypeInManager(overlayNetworkName, types.Overlay) {
-			return "", fmt.Errorf("network %s does not match type %q in manager", overlayNetworkName, types.Overlay)
+
+		if len(globalBGPNetworkName) != 0 {
```

Comment:
can you judge if returned network name is empty inside `getNetworkByNodeNameIndexer`, then we can save another 3 lines.

Reply:
It might be worthless, because you always need to check if the network exist and return an error if it's not exist, whether you return network name is empty inside `getNetworkByNodeNameIndexer` or not.